spark2/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,368 @@
+
+    LOG.info("Processing data from {} to {}", startOffset, endOffset);
+    List<FileScanTask> tasks = pendingBatches.stream()
+        .flatMap(batch -> batch.tasks().stream())
+        .collect(Collectors.toList());
+    CloseableIterable<FileScanTask> splitTasks = TableScanUtil.splitFiles(CloseableIterable.withNoopClose(tasks),
+        splitSize);
+    List<CombinedScanTask> combinedScanTasks = Lists.newArrayList(
+        TableScanUtil.planTasks(splitTasks, splitSize, splitLookback, splitOpenFileCost));
+
+    if (readUsingBatch == null) {
+      this.readUsingBatch = checkEnableBatchRead(combinedScanTasks);
+    }
+
+    return combinedScanTasks;
+  }
+
+  /**
+   * Used to calculate start offset. If the startSnapshotId has a value, start the construction
+   * from the specified snapshot, otherwise, start the construction from the beginning.
+   *
+   * @return The start offset to scan from.
+   */
+  private StreamingOffset calculateStartingOffset() {
+    StreamingOffset startingOffset;
+    if (startSnapshotId != null) {
+      startingOffset = new StreamingOffset(startSnapshotId, 0, false);
+    } else {
+      List<Long> snapshotIds = SnapshotUtil.currentAncestors(table);
+      if (snapshotIds.isEmpty()) {
+        // there's no snapshot currently.
+        startingOffset = StreamingOffset.START_OFFSET;
+      } else {
+        startingOffset = new StreamingOffset(Iterables.getLast(snapshotIds), 0, true);
+      }
+    }
+
+    return startingOffset;
+  }
+
+  /**
+   * Used to calculate end offset.
+   *
+   * @param start The start offset to scan from
+   * @return The end offset to scan to
+   */
+  private StreamingOffset calculateEndOffset(StreamingOffset start) {
+    if (start.equals(StreamingOffset.START_OFFSET)) {
+      return StreamingOffset.START_OFFSET;
+    }
+
+    // Spark will invoke setOffsetRange more than once. If this is already calculated, use the cached one to avoid
+    // calculating again.
+    if (cachedPendingBatches == null || CollectionUtils.isEmpty(cachedPendingBatches.second()) ||
+        !cachedPendingBatches.first().equals(start)) {
+      this.cachedPendingBatches = Pair.of(start, getChangesWithRateLimit(start, maxSizePerBatch));
+    }
+
+    List<MicroBatch> batches = cachedPendingBatches.second();
+    MicroBatch lastBatch = Iterables.getLast(batches, null);
+
+    if (lastBatch == null) {
+      return start;
+    } else {
+      boolean isStarting = lastBatch.snapshotId() == start.snapshotId() && start.shouldScanAllFiles();
+      return new StreamingOffset(lastBatch.snapshotId(), lastBatch.endFileIndex(), isStarting);
+    }
+  }
+
+  /**
+   * Streaming Read control is performed by changing the startOffset and maxSize.
+   *
+   * @param startOffset The start offset to scan from
+   * @param maxSize     The maximum size of Bytes can calculate how many batches
+   * @return MicroBatch of list
+   */
+  @VisibleForTesting
+  @SuppressWarnings("checkstyle:HiddenField")
```

Comment:
why suppress this warning?

Reply:
```
Task :iceberg-spark2:checkstyleMain FAILED
[ant:checkstyle] [ERROR] /opt/sourcecode/iceberg-src/spark2/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java:281:60: 'startOffset' 隐藏属性。 [HiddenField]
```

Reply:
> why suppress this warning?

```
Task :iceberg-spark2:checkstyleMain FAILED
[ant:checkstyle] [ERROR] /opt/sourcecode/iceberg-src/spark2/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java:281:60: 'startOffset' 隐藏属性。 [HiddenField]
```

Reply:
I think generally we just change the field visibility